### PR TITLE
Add support for optional catch binding syntax

### LINKED ISSUE #5210: optional catch binding
[This proposal](https://github.com/tc39/proposal-optional-catch-binding) is now in stage 4, and is being used in node.js core.

### PR DESCRIPTION
**Pull request by zenparsing**
Adds support for [optional catch binding syntax](https://tc39.github.io/proposal-optional-catch-binding), now at stage 4 and implemented in V8, SM, and JSC.

This is my first PR; I'll appreciate any and all feedback!

Resolves #5210

### CODE REVIEW

**Comment by pleath** [comment]
Nice! Thanks, @zenparsing . Two things that would be good to confirm are that heap enumeration and F12 debugging can handle this syntax. There are heap enum tests in the full repo and tests in DebuggerCommon in the core repo that you can use as a model for testing F12 enumeration of locals.

File: test/EH/optional-catch-binding.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  },
+  {
+    name: "Error are correctly thrown from catch",
+    body: function() {
+      class Err {}
+      assert.throws(function() {
+        try {
+          throw new Error();
+        } catch {
+          throw new Err();
```
Comment (akroshg) [inline test/EH/optional-catch-binding.js]: is there a rethrow syntax as well?

File: lib/Parser/ptree.cpp [hunk 2]
```diff
@@ -578,6 +578,7 @@ ParseNodeTry::ParseNodeTry(OpCode nop, charcount_t ichMin, charcount_t ichLim)
 ParseNodeCatch::ParseNodeCatch(OpCode nop, charcount_t ichMin, charcount_t ichLim)
     : ParseNodeStmt(nop, ichMin, ichLim)
 {
+    this->pnodeParam = nullptr;
```
Comment (akroshg) [inline lib/Parser/ptree.cpp]: hmmm interesting. Could you initialize this at initialization list of the constructor?

**Comment by akroshg** [comment]
>         Indent(indentAmt);

do you need to update this? try running your scenario with -trace:parse and see everything as expected.

---
Refers to: lib/Parser/Parse.cpp:13730 in 54695a0. [](commit_id = 54695a083631de91bd77d53c197943d2494e3ef2, deletion_comment = False)

**Comment by akroshg** [comment]
Thanks for mentioning that. Also I'd like few test to be added with new syntax which are under debugger and heap-enum.

---
In reply to: [397101019](https://github.com/Microsoft/ChakraCore/pull/5310#issuecomment-397101019) [](ancestors = 397101019)

**Comment by aneeshdk** [comment]
Are we not putting this under a flag?

**Comment by aneeshdk** [comment]
Can you add couple more test cases with eval, var leaking in catch scope, with scope in catch block and function definitions inside catch block?

**Comment by akroshg** [comment]
@aneeshdk - we should have flag to on and off this feature. But I'd vote for having this enable by-default ON.

**Comment by akroshg** [comment]
@zenparsing are there any test262 tests for this feature and do we know how many pass/fail?

File: test/EH/optional-catch-binding.js (hunk 1 above)
Comment (kfarnung) [reply test/EH/optional-catch-binding.js]: Taking a look at the proposal it doesn't look like it.  Seems like having `throw;` would have been interesting (like C# does), but it seems like if you want to rethrow then you'll need the parameter binding.

Comment (fatcerberus) [reply test/EH/optional-catch-binding.js]: Personally I like that you still have to be explicit if you want to rethrow, I always found `throw;` in C++ to be kind of obtuse.

**Comment by kfarnung** [comment]
@zenparsing @akroshg @aneeshdk What is left to get this landed?

**Comment by zenparsing** [comment]
@kfarnung I've updated tests and rebased, but I have not yet implemented a flag for this feature.

**Comment by zenparsing** [comment]
Do we still want an always-true flag for this? It's been unflagged in v8 for a while now.

**Comment by pleath** [comment]
If the syntax is sure to stay in the spec, I think we can do without the flag.

**Comment by thw0rted** [comment]
Is this patch never going to reach production because Edge is switching to a Chromium backend now?  [This SO question](https://stackoverflow.com/questions/21624456/can-i-use-a-try-catch-in-javascript-without-specifying-the-catch-argument-identi) leaves devs with the impression that Edge should either support the syntax already, or is going to at some point in the future.  Is that still true?

**Pull request merged**